### PR TITLE
[tt-train] Forward pass for CrossEntropyLoss

### PR DESCRIPTION
**Pull request by vmelnykovTT**
### Ticket
[[Feature Request] Cross entropy loss forward and backward functions](https://github.com/tenstorrent/tt-metal/issues/14113)

### Problem description
Currently, a cross entropy loss operation does not exist.
Instead, we use a composite operation to perform both the forward and backward computations manually.

### What's changed
In this PR, the `cross_entropy_fw` operation is added.
The `cross_entropy_bw` operation will be implemented in a separate PR

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/kernels/compute/cross_entropy_fw_kernel.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+                sub_bcast_cols_init_short(cb_input, cb_max_value_after_reduction);
+                sub_tiles_bcast_cols(
+                    cb_input,
+                    cb_max_value_after_reduction,
+                    /* tile idx */ col,
+                    /* tile idx */ 0,
+                    /* reg tile idx */ working_register);
+
+                mul_bcast_cols_init_short(cb_target, cb_exp_sum_after_reduction);
+                mul_tiles_bcast_cols(
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/kernels/compute/cross_entropy_fw_kernel.cpp:246]: we can multiply by target once (in this code it's twice)

File: tt-train/tests/ops/cross_entropy_fw_op_test.cpp [hunk 2]
```diff
@@ -0,0 +1,244 @@
+
```
Comment (rfurko-tt) [inline tt-train/tests/ops/cross_entropy_fw_op_test.cpp:1]: remove empty line

**Review by rfurko-tt — COMMENTED** [review]
Overall looks good. Great! 

Unfortunately, wasn't able to find bugs (why it does not work during first run after `tt-smi -r`)

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/cross_entropy_fw.cpp [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#include "cross_entropy_fw.hpp"
+
+#include "device/cross_entropy_fw_device_operation.hpp"
+
+namespace ttml::metal::ops::cross_entropy_fw {
+
+ttnn::Tensor CrossEntropyForwardOperation::invoke(const ttnn::Tensor& input_tensor, const ttnn::Tensor& target_tensor) {
+    auto result /*ttnn::Tensor*/ = ttnn::prim::ttml_cross_entropy_fw(
```
[the remark this answers is not in the document]
Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/cross_entropy_fw/cross_entropy_fw.cpp:12]: Yeah, the function return value also indicates the type of the result

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+        bfloat16_single_tile_size_bytes,
+        kNumOutputBeforeReductionTiles);
+
+    auto cb_output = create_circular_buffer(
+        program,
+        all_cores,
+        kOutputCbIndex,
+        data_format,
+        bfloat16_single_tile_size_bytes,
+        2 * block_size);  // create twice block_size variable
```
[the remark this answers is not in the document]
Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp:322]: I only need one tile here, since I get the result and write it to the output once

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/kernels/compute/cross_entropy_fw_kernel.cpp (hunk 1 above)
Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/kernels/compute/cross_entropy_fw_kernel.cpp]: Changed the calculation approach and completely avoided broadcasting

File: tt-train/tests/ops/cross_entropy_fw_op_test.cpp [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+TEST_F(CrossEntropyForwardTest, CrossEntropyForward_Huge_Forward) {
+    using namespace ttml;
+
+    const uint32_t N = 64U, C = 1U, H = 32U, W = 128000U;
+    const auto shape = ttnn::SmallVector<size_t>{N, C, H, W};
+
+    std::random_device rd;
+    std::mt19937 gen(rd());  // or fixed seed: std::mt19937 gen(42);
```
Comment (rfurko-tt) [inline tt-train/tests/ops/cross_entropy_fw_op_test.cpp:253]: change to fixed seed please

**Review by rfurko-tt — COMMENTED** [review]
Looks good overall! Please clean it a bit more and ready to go 

@dmakoviichuk-tt could you please take a look as well?

**Comment by rfurko-tt** [comment]
Please also update PR description

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rfurko-tt) [reply tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp]: you want to have two, while writing one tile you want to start preparing another one

**Comment by rfurko-tt** [comment]
it would be also great to update cross entropy operation and MNIST demo

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_device_operation.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    const auto& input_logical_shape = input_tensor.get_logical_shape();
+    auto program_factory = select_program_factory(args, tensor_args);
+    tt::tt_metal::operation::Hash hash = tt::tt_metal::operation::hash_operation<CrossEntropyForwardDeviceOperation>(
+        args, program_factory.index(), input_tensor.dtype(), input_logical_shape);
+
+    return hash;
+}
+
+std::
+    tuple<CrossEntropyForwardDeviceOperation::operation_attributes_t, CrossEntropyForwardDeviceOperation::tensor_args_t>
```
[the remark this answers is not in the document]
Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_device_operation.cpp:119]: This formatting happens automatically when I save or run clang-format, so I’m not manually formatting it that way

File: tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+    uint32_t Ht = padded_tensor_shape[-2] / tt::constants::TILE_HEIGHT;
+    uint32_t NC = padded_tensor_shape[0] * padded_tensor_shape[1];
+    uint32_t total_rows_to_process = NC * Ht;
+
+    // get size of target indexes inner dimension
+    uint32_t target_indexes_inner_dim_size = target.get_logical_shape()[-1] * target.element_size();
+    // read target indexes by pages(32 indexes in page)
+    uint32_t uint32_read_page_size = tt::datum_size(tt::DataFormat::UInt32) * kPageElementsNumber;
+    // tiled height of target indexes
+    uint32_t tiled_H =
```
[the remark this answers is not in the document]
Comment (vmelnykovTT) [reply tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp:239]: I don't clearly understand what do you mean?

[the remark this answers is not in the document]
Comment (romaf5) [reply tt-train/sources/ttml/metal/ops/cross_entropy_fw/device/cross_entropy_fw_program_factory.cpp:238]: Ht = tiled_H

**Review by dmakoviichuk-tt — APPROVED** [review]
